sympy/solvers/ode/single.py

```diff
@@ -1975,5 +2045,654 @@ def _get_general_solution(self, *, simplify_flag: bool = True):
+        return retdict
+
+    def _solve_undetermined_coefficients(self,match):
+        r"""
+        Helper function for the method of undetermined coefficients.
+
+        See the
+        :py:meth:`~sympy.solvers.ode.single.NthConstCoeffUndet`
+        docstring for more information on this method.
+
+        The parameter ``match`` should be a dictionary that has the following
+        keys:
+
+        ``list``
+        A list of solutions to the homogeneous equation.
+
+        ``sol``
+        The general solution,.
+
+        ``trialset``
+        The set of trial functions as returned by
+        ``_undetermined_coefficients_match()['trialset']``.
+
+        """
+        eq = self.ode_problem.eq_high_order_free
+        x = self.ode_problem.sym
+        f = self.ode_problem.func.func
+        order = self.ode_problem.order
+        r = match
+        coeffs = numbered_symbols('a', cls=Dummy)
+        coefflist = []
+        gensols = r['list']
+        gsol = r['sol']
+        trialset = r['trialset']
+        if len(gensols) != order:
+            raise NotImplementedError("Cannot find " + str(order) +
+            " solutions to the homogeneous equation necessary to apply" +
+            " undetermined coefficients to " + str(eq) +
+            " (number of terms != order)")
+
+        trialfunc = 0
+        for i in trialset:
+            c = next(coeffs)
+            coefflist.append(c)
+            trialfunc += c*i
+
+        eqs = sub_func_doit(eq, f(x), trialfunc)
+
+        coeffsdict = dict(list(zip(trialset, [0]*(len(trialset) + 1))))
+
+        eqs = _mexpand(eqs)
+
+        for i in Add.make_args(eqs):
+            s = separatevars(i, dict=True, symbols=[x])
+            if coeffsdict.get(s[x]):
+                coeffsdict[s[x]] += s['coeff']
+            else:
+                coeffsdict[s[x]] = s['coeff']
+
+        coeffvals = solve(list(coeffsdict.values()), coefflist)
+
+        if not coeffvals:
+            raise NotImplementedError(
+                "Could not solve `%s` using the "
+                "method of undetermined coefficients "
+                "(unable to solve for coefficients)." % eq)
+
+        psol = trialfunc.subs(coeffvals)
+
+        return Eq(f(x), gsol.rhs + psol)
+
+    def _get_general_solution(self, *, simplify_flag: bool = True):
+        eq = self.ode_problem.eq_high_order_free
+        f = self.ode_problem.func.func
+        x = self.ode_problem.sym
+        homogen_instance = NthConstCoeffHomogen(SingleODEProblem(eq, f(x), x))
+        gensols, collectterms = homogen_instance._get_sols(self.r)
+        # A generator of constants
+        constants = self.ode_problem.get_numbered_constants(num=len(gensols))
+        gsol = Add(*[i*j for (i, j) in zip(constants, gensols)])
+        gsol = Eq(f(x), gsol)
+        self.r.update({'list': gensols, 'sol': gsol, 'simpliy_flag': simplify_flag})
+        gsol = self._solve_undetermined_coefficients(self.r)
+        if simplify_flag:
```

Comment:
Also, this raises an exception  
```
File "/home/runner/work/sympy/sympy/sympy/solvers/ode/single.py", line 2691, in _get_general_solution
    gsol = self._solve_undetermined_coefficients(self.r)
 File "/home/runner/work/sympy/sympy/sympy/solvers/ode/single.py", line 2668, in _solve_undetermined_coefficients
    coeffvals = solve(list(coeffsdict.values()), coefflist)
    raise GeneratorsNeeded(
sympy.polys.polyerrors.GeneratorsNeeded: can't initialize from 'dict' without generators
```
whereas when I used inheritance, it was working fine.